### ISSUE TITLE
Spacemacs/Emacs inspired web browser

### ISSUE THREAD

**Issue by bendiksolheim** [mikaelbr/open-source-ideas#9]
## Project description

As a computer nerd, I spend a lot of time in my web broser. I constantly try to make this a more pleasant experience, but I can't seem to become completely happy. What annoys me the most is the lack og good hotkeys. Hotkeys are mostly modelled after the well known [modifier] + [key] system, which is quite limiting when you have experienced the power of Vi or Emacs.

A project which gets quite close to this is [Vimperator](https://github.com/vimperator/vimperator-labs). It is heavily inspired by Vim, with normal mode and insert mode, known hotkeys, searching and so on. The main problem with Vimperator is that it is an addon, and sort of feels "non-native". It is certainly a very good alternative, but there are so many more interesting ideas I would like to explore:

- Remove the notion of tabs. You instead have a number of _buffers_, each containing one website. The browser initially has _one window_, this can show _one buffer_. You can swap buffers whenever you like (which would be the same as chaning tabs). A window can be split in half horizonally of vertically to create more windows, displaying multiple buffers at the same time.
- Sort buffers into "projects". With this you could sort your buffers into work buffers and private buffers, and load the set of buffers relevant to what you are doing right now.
- Extremely lightweight GUI. With complex hotkeys, you can hide away a lot of the stuff that is normally shown in a GUI. Hotkeys are often critisized for being hard to discover, but as Spacemacs has shown this does not neccesarily need to be true. You can also show most stuff you need to show (such as state of the history stack, current buffer URL and so on) in a line at the top or bottom of the window, completely removing the need for large interface icons.

These are just a few of the ideas for such a browser. The main idea is to make browsing feel more like editing in Emacs (or Vi, as a lot of the ideas exists there as well). It will certainly not be a mainstream browser, but the hope is that it will give "advanced users" a more pleasant experience when browsing the web.

This is very much inspired by the editor [Spacemacs](http://spacemacs.org/) (which really is just a heavily configured Emacs distribution), so a few of the concepts might make less sence if you haven't used it.

## Relevant Technology

Making a web browser completely from scratch is a more or less impossible task today. It would probably be a good idea to base the browser on Webkit (or similar). When it comes to language, that is a though one. The first thing that comes to mind is using a technology such as [Electron](http://electron.atom.io/) since that already is integrated with Webkit and is quite widely accessible as it is Javascript based. However, one needs to be certain that this is a task Javascript can handle.

The most important thing, I believe, it that the browser should run on "all" major desktop operating systems. We are already limiting the user base to advanced users, so it would be bad to limit it even further by supporting only on desktop operating system.

## Who is this for

This project will definitely need experienced developers on the board. Some of the tasks required by a web browser (security, to name one) are really complex. But there will definitely be room for more novice or beginner programmers as well.

This idea is probably better suited for a team, and not just a single person (but please prove me wrong!).

**Comment by FredrikAugust** [comment]
Sounds a bit like vimb :)

**Comment by tbodt** [comment]
Sounds pretty cool. The other day I tried to close a tab using `:q`.

**Comment by bendiksolheim** [comment]
Do you know if vimb is being actively developed @FredrikAugust ? I had a look at it a before I made this issue, but I could not find a link to the source back then so I was a bit unsure if it was still actively developed. Poked around a bit more just now, and the last commit was actually just 16 days ago. Have you tried it?

**Comment by FredrikAugust** [comment]
@bendiksolheim I used it on and off for about a month now, and it seems to be pretty complete. Only thing I was missing was flexbox support, which is kind of a deal-breaker for me, as I use them a lot for web dev. As far as development goes I have no clue.

**Comment by tbodt** [comment]
Likely good idea for implementation would be to use Electron.

**Comment by Thrilleratplay** [comment]
@bendiksolheim Have you tried [Vimium](https://github.com/philc/vimium)?

**Comment by bendiksolheim** [comment]
@Thrilleratplay : yeah, I used it back when I used Chrome :) It is about the same as Vimperator for Firefox, and as far as I remember also suffers a bit from being an extension.

I haven't had the time to try out Vimb yet, but I definitely will.

**Comment by azer** [comment]
[kaktüs](http://github.com/kaktus/kaktus) aims to achieve this, you can give it a try. You can open 100 tabs with it and filter everything (tabs, bookmarks and history) by keywords. Feel free to contribute it. 

Previously in Linux, I had much more minimal solution with [delicious surf](http://github.com/azer/delicious-surf) but it needed a tiling window manager like Xmonad.

**Comment by ethanhs** [comment]
Well, there is [Qutebrowser](https://github.com/The-Compiler/qutebrowser), which does something like this for VIM.

**Comment by edrex** [comment]
qutebrowser is the closest thing I've found.

IMO browser extension APIs are too limited to provide a complete keyboard-driven UI (both on the display and control end) - you need a dedicated browser like qutebrowser to provide a cohesive experience.

These niche projects need support, both in testing and donations (@The-Compiler [ran an indiegogo](https://www.kickstarter.com/projects/the-compiler/qutebrowser-v10-with-per-domain-settings) for qutebrowser and you can [donate](https://github.com/qutebrowser/qutebrowser#donating) directly) or they will fizzle out, so if you have the resources and motivation, consider contributing.

**Comment by bendiksolheim** [comment]
I have ended up giving qutebrowser a go now that Firefox has decided to remove support for XUL, and I have to admit that it has a lot of the things I am looking for in a browser. People coming from Vimperator should feel right at home! It would be nice if it could be integrated with my password manager, but I can live with copy pasting passwords.

**Comment by edrex** [comment]
I use password-store (the "pass" command) to manage my passwords, and there is a qutebrowser [userscript](https://github.com/qutebrowser/qutebrowser/blob/master/misc/userscripts/password_fill) ([background](https://github.com/qutebrowser/qutebrowser/pull/873)) which can auto-fill forms from that. I haven't tried it but it's on my list.

**Comment by koppor** [comment]
Related work:

- [webmacs](https://github.com/parkouss/webmacs) - actively maintained
- [conkeror](http://conkeror.org/) - not maintained ([repository](https://github.com/retroj/conkeror))
- [emacs browsin plugins](https://github.com/emacs-tw/awesome-emacs#browser)

**Comment by edrex** [comment]
interesting to see webmacs also using pyqt/qtwebengine. the more downstream projects use that platform, the better blazed the trail will be for other alt browsers.
